### PR TITLE
Port to Windows(1)

### PR DESCRIPTION
**Pull request by taozuhong**
refactor fnmatch with g_pattern_match_simple

### CODE REVIEW

File: src/as-component.c [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
 #include <glib.h>
 #include <glib-object.h>
-#include <fnmatch.h>
+#include <glib/gpattern.h>
```
[the remark this answers is not in the document]
Comment (taozuhong) [reply src/as-component.c:26]: u r right, I'll update it.

File: tools/appstreamcli.c [hunk 2]
```diff
@@ -1089,7 +1089,7 @@ as_client_run_compose (const gchar *command, char **argv, int argc)
 	for (gint i = 2; i < argc; i++)
 		asc_argv[i-1] = argv[i];
 
-	return execv(ascompose_exe, (gchar * const*) asc_argv);
```
Comment (ximion) [inline tools/appstreamcli.c:1092]: This replacement is wrong, we must call `execv` here to replace the current process with the new one. If Windows doesn't have a compatibility interface for this, I'd rather like a separate codepath here for Windows that uses spawn_sync/posix_spawn and doesn't replace the parent (so the original behavior on non-Windows is preserved).

**Review by ximion — CHANGES_REQUESTED** [review]

File: tools/appstreamcli.c [hunk 3]
```diff
@@ -1068,7 +1068,7 @@ static int
 as_client_run_compose (const gchar *command, char **argv, int argc)
 {
 	const gchar *ascompose_exe = LIBEXECDIR "/appstreamcli-compose";
-	g_autofree const gchar **asc_argv = NULL;
```
Comment (ximion) [inline tools/appstreamcli.c:1071]: Why did you remove the `const`? This will most likely break the following code when `ascompose_exe` and other constant strings are assigned (we really just want to allocate the container here, not the string contents).

Comment (taozuhong) [reply tools/appstreamcli.c:1071]: Ah, the `const` will break up with compilation.

File: src/as-distro-extras.c [hunk 4]
```diff
@@ -62,18 +62,18 @@ static gboolean
 directory_is_empty (const gchar *dirname)
 {
 	gint n = 0;
-	struct dirent *d;
-	DIR *dir = opendir (dirname);
```
[the remark this answers is not in the document]
Comment (taozuhong) [reply src/as-distro-extras.c:66]: Maybe have, but GDir is good for portability

[the remark this answers is not in the document]
Comment (taozuhong) [reply src/as-distro-extras.c:66]: thanks

**Comment by elmarco** [comment]
@taozuhong are you working on this?

File: src/as-distro-extras.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (elmarco) [reply src/as-distro-extras.c:66]: @ximion there is no reason for gdir to be much slower than DIR, it's a portable wrapper: https://gitlab.gnome.org/GNOME/glib/-/blob/main/glib/gdir.c

**Comment by elmarco** [comment]
ok, I am taking it over in #510. You may close this PR.
